oracle/oracle.go

```diff
@@ -197,6 +199,36 @@ func (or *Oracle) AdvanceStateToNextSlot(fullBlock *FullBlock) (uint64, error) {
 	return processedSlot, nil
 }
 
+// Unsubscribes validators that are not active in ethereum. Shares their pending rewards to the pool
+// TODO: there are multiple validator statuses that should be considered as "not active" and should be unsubscribed.
+func (or *Oracle) ValidatorCleanup(blockHeader *v1.BeaconBlockHeader) error {
+	or.mutex.Lock()
+	defer or.mutex.Unlock()
+
+	// Only cleanup if we're past the cleanup slot fork
+	if uint64(blockHeader.Header.Message.Slot) >= MainnetCleanupValidatorsSlot {
+		// Get the latest validator information
+		validatorInfo := or.validators.GetValidators()
+
+		// Iterate over all validators. If two or more  validators exit or get slashed in the same slot,
+		// this cleanup will eventually set both of their pending rewards to 0 and share them among the pool
+		for _, validator := range validatorInfo {
+			// If a validator is subscribed but not active onchain, we have to unsubscribe it and treat it as a ban:
+			// this means setting the validator rewards to 0 and sharing them among the pool
+			if !validator.Status.IsActive() && or.isSubscribed(uint64(validator.Index)) {
+				or.advanceStateMachine(uint64(validator.Index), Unsubscribe)
```

Comment:
not sure if we should use the state machine. Up until now the event Unsubscribe was only triggered by real subscriptions (contract events) so this may be confusing.

Reply:
you're right. Logs would be confusing if we use state machine here, but its functionality is perfect in this usecase